SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,716 @@
+      )
+
+      if !isClosurePassedTrivially {
+        continue
+      }
+    }
+
+    let onlyHaveThinToThickClosure =
+      closure is ThinToThickFunctionInst && !haveUsedReabstraction
+
+    guard let closureParamInfo = pai.operandConventions[parameter: use.index] else {
+      fatalError("Parameter info not found for operand: \(use)!")
+    }
+
+    if (closureParamInfo.convention.isGuaranteed || isClosurePassedTrivially)
+      && !onlyHaveThinToThickClosure
+    {
+      continue
+    }
+
+    // Functions with a readnone, readonly or releasenone effect and a
+    // nontrivial context cannot be specialized. Inserting a release in such
+    // a function results in miscompilation after other optimizations.
+    // For now, the specialization is disabled.
+    //
+    // TODO: A @noescape closure should never be converted to an @owned
+    // argument regardless of the function's effect attribute.
+    if !callee.effectAllowsSpecialization && !onlyHaveThinToThickClosure {
+      continue
+    }
+
+    // Avoid an infinite specialization loop caused by repeated runs of
+    // ClosureSpecializer and CapturePropagation.
+    // CapturePropagation propagates constant function-literals. Such
+    // function specializations can then be optimized again by the
+    // ClosureSpecializer and so on.
+    // This happens if a closure argument is called _and_ referenced in
+    // another closure, which is passed to a recursive call. E.g.
+    //
+    // func foo(_ c: @escaping () -> ()) {
+    //   c()
+    //   foo({ c() })
+    // }
+    //
+    // A limit of 2 is good enough and will not be exceed in "regular"
+    // optimization scenarios.
+    let closureCallee =
+      closure is PartialApplyInst
+      ? (closure as! PartialApplyInst).referencedFunction!
+      : (closure as! ThinToThickFunctionInst).referencedFunction!
+
+    if closureCallee.specializationLevel > specializationLevelLimit {
+      continue
+    }
+
+    if callSiteMap[pai] == nil {
+      callSiteMap.insert(key: pai, value: CallSite(applySite: pai))
+    }
+
+    intermediateClosureArgDescriptors.append(
+      IntermediateClosureArgDescriptor(
+        callSite: pai,
+        closureArgumentIndex: closureArgumentIndex,
+        parameterInfo: closureParamInfo,
+        reachableExitBBs: Array(callee.blocks.filter { $0.isReachableExitBlock })
+      )
+    )
+  }
+
+  finalizeCallSites(
+    context,
+    for: closure,
+    callSiteMap: &callSiteMap,
+    lifeRangeEndpoints: lifeRangeEndpoints,
+    intermediateClosureArgDescs: intermediateClosureArgDescriptors
+  )
+
+  return false
+}
+
+/// Handles conversion and reabstraction uses of the supported closure.
+///
+/// - Parameters:
+///   - context: `FunctionPassContext` instance
+///   - for: A closure against which callees may be specialized.
+/// - Returns:
+///   - remainingUses: Any uses not handled in this function.
+///   - lifeRangeEndpoints: Instructions that comprise the possible life-range of the closure.
```

Comment:
The name `lifeRangeEndpoints ` is wrong. It's a list of _all_ users, not just the life range end points